### ISSUE TITLE
Support Interactive Rebases

### ISSUE THREAD

**Issue by Kurt-von-Laven** [desktop/desktop#12354]
### Describe the feature or problem you’d like to solve

I would like to be able to do interactive rebases in GitHub Desktop without using the command line.

### Proposed solution

Interactive rebases are one of the crowning features of git. They offer enormous flexibility to craft a revision history to a wide variety of specifications, and can be used to remedy almost all of the common situations engineers face after having made a mistake with git.

### Additional context

This is the git/GitHub workflow I was taught by Linux kernel developers, the original purpose for which git was designed. It has a bit of a learning curve, but produces a commit history of the highest quality when applied correctly. With a bit of practice my experience has been that it actually accelerates even my initial authoring of correct software as opposed to slowing me down or distracting from feature development. The practice of decomposing my changes into correct commits and authoring helpful commit messages often reveals my own mistakes to me. I don't recommend this workflow, however, if your goal is to hastily churn out buggy code; nor do I recommend attempting to transition to it during a period of unusual organizational stress. When the commits are clean and coherent, it facilitates code review, maintenance (especially when an engineer leaves the organization), debugging with `git bisect`, the use of `git blame` to offer context, etc. I would love to see interactive rebases supported by GitHub Desktop so that engineers who are less familiar with git and the command line can devote less of their focus to ramping up on the git CLI. In the meantime, I will encourage use of [GitLens in VSCode](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens#interactive-rebase-editor), but that has the drawback of requiring people to either use two different editors or leave their editor of choice. I kept the explanations of the commands as a reminder of how much newcomers to git are already being asked to learn at once.

1. `git checkout main` (Switch to the main branch if you aren't already on it.)
2. `git pull` (Update the main branch with any changes to the main branch on GitHub.)
3. `git checkout -b <branch_name>` (Make a new branch and switch to it.)
4. Make the pertinent changes to your task. (It's fine to fix other things along the way within reason if desired, but put them in separate commits.)
5. `git add -p <modified_file>` (Select the appropriate hunks to stage from each file you made changes to, repeating as needed. This process is invaluable for self review as we are psychologically more likely to detect our own mistakes when presented with them in a different format.)
6. `git commit`  (Repeat steps 5-6 once per logical coherent describable unit of work; all of the changes within one commit should relate to one another.)
7. `git push --set-upstream origin <branch_name>` (Copy your branch to GitHub. Aim for fewer than 100 to 300 lines per pull request, which is all the commits you push in one unit for review.)
8. Open a pull request to merge your branch to main at GitHub. (The output of git push contains a link to do this.)
9. Address review feedback, if any.
10. `git add -p <modified_file>` (This is the same as step 5, but the changes will be relative to the changes you previously made.)
11. `git commit` (Add the `--amend` flag if there is only one commit in your pull request to squash all your changes together and avoid polluting the revision history with a bunch of commits that say "addressed code review feedback.")
12. `git rebase -i main` (If there is more than one commit in your pull request, you can use this command to fixup specific commits in response to your review feedback.)
13. `git push --force-with-lease` (Overwrite your branch on GitHub with the latest and greatest version of your branch.)
14. Once your pull request is approved, rebase and merge your branch into main using the green button at the bottom, and delete the branch on GitHub.
15. `git checkout main` (Switch to the main branch.)
16. `git pull` (Update the main branch with the contents of your recently merged pull request and any other changes that may have landed in the meantime.)
17. `git checkout <branch_name>` (Switch to the feature branch.)
18. `git rebase main` (Rebase the feature branch to ensure that you merged what you expected to; this should succeed without intervention.)
19. `git checkout main` (Switch back to main.)
20. `git branch -d <branch_name>` (Delete the local branch now that it's merged and rebased.)

**Comment by sergiou87** [comment]
We're not planning to add "interactive rebases" directly, but in different operations like you've seen in some places: #12235 or #12299 

Interactive rebase is a quite heavy process, so we're trying to simplify it aiming to solve specific use cases 😄 

Do you think that would work for you?

**Comment by Kurt-von-Laven** [comment]
Thank you for communicating your plans. I realize tools are not tacos :taco:, and hence can't please everyone. The main barrier to our adopting GitHub Desktop as our primary GUI tool for git is that we have no way to squash new changes on a branch into a specific prior commit, which is important for addressing code review feedback on PRs that have more than one commit (in our case, virtually all of them). Our use case could be satisfied is if we could commit staged changes directly to a particular commit, combine particular commits that are not adjacent, or (best of all) edit a particular commit directly. It is theoretically possible to achieve our workflow by reordering commits, squashing, and then reordering back. The issue with that is it's (a) a lot of extra clicks and (b) forces you to resolve any merge conflicts twice or even four times instead of once.

**Comment by Kurt-von-Laven** [comment]
This feature could be implemented using [git commit --fixup](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt) and [git commit --squash](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---squashltcommitgt).

**Comment by sergiou87** [comment]
Sorry I missed your reply 😓 

> Our use case could be satisfied is if we could commit staged changes directly to a particular commit, **combine particular commits that are not adjacent**

Isn't #12235 what you're describing there? If you install the [latest beta](https://github.com/desktop/desktop#beta-channel), you can basically drag any commit onto any other commit to squash them

> edit a particular commit directly

We also added this in #12353 but only applies to the last commit as of today 😅

**Comment by Kurt-von-Laven** [comment]
> Sorry I missed your reply 

No worries.

> Isn't #12235 what you're describing there? If you install the [latest beta](https://github.com/desktop/desktop#beta-channel), you can basically drag any commit onto any other commit to squash them

Oh sweet! Indeed it is. Apologies for overlooking that issue. I see now that you already sent that above.

> We also added this in #12353 but only applies to the last commit as of today

Yeah, this one I was aware of. It sounds like GitHub Desktop handles our workflow with a maximum of one unnecessary conflict resolution rather than three at this point then, which is great news. I think the only remaining request to get it down to zero unnecessary conflict resolutions would be the ability to edit a particular commit directly, meaning to temporarily return the branch to a particular prior state where one could make modifications, use the amend functionality you already offer, and then continue back to the present. As an added benefit, this would allow people to examine the state of the branch at a previous commit regardless of whether they desired to change it. I imagine this would be particularly desirable for a novice, who may find reading a diff very challenging initially. Right now, I think one could achieve this in GitHub Desktop by limiting themselves to one commit per branch, and then rebasing each branch in turn on its parent, but that is fairly error-prone and requires many extra clicks.

**Comment by JKRhb** [comment]
@sergiou87 The reordering and squashing features are very awesome, thank you for adding them! I think at the moment there are mainly two features missing to get a very good rebase experience using GitHub Desktop:

1. Being able to drop (i. e. delete) commits from the history
2. Being able to change a commit message without squashing

Are you planning on adding in these as well?

**Comment by Kurt-von-Laven** [comment]
@JKRhb, they already added two if I understand you correctly. There is a checkbox to amend a commit.

**Comment by JKRhb** [comment]
> @JKRhb, they already added two if I understand you correctly. There is a checkbox to amend a commit.

Hmm, but this way you can only rename the latest commit when it isn't pushed already if I am not mistaken? I meant that you can click on any commit and choose something like "Update commit message" and then effectively perform an interactive rebase with the `reword` operation.

**Comment by Kurt-von-Laven** [comment]
You can rename it locally even if it is already pushed and then force push the branch. This is generally a good idea if it's your own branch and a bad one if it's shared (e.g., master/main). To reword a commit, presently you would have to reorder them (e.g., via drag and drop), amend it, and then reorder them back. I agree that it would be ideal to be able to reword a commit without this level of overhead.

**Comment by sergiou87** [comment]
Thanks for your feedback @JKRhb!! ❤️ 

I'm afraid right now those features are not in our roadmap. I recommend to file issues for both features (if they don't exist already). Getting traction on those (via 👍  reactions or comments) will make them more likely to happen, as that would mean there is an interest/need on them.

**Comment by 0xdevalias** [comment]
RE: https://github.com/desktop/desktop/issues/12354#issuecomment-854266676

> Our use case could be satisfied is if we could commit staged changes directly to a particular commit, combine particular commits that are not adjacent, or (best of all) edit a particular commit directly. It is theoretically possible to achieve our workflow by reordering commits, squashing, and then reordering back. The issue with that is it's (a) a lot of extra clicks and (b) forces you to resolve any merge conflicts twice or even four times instead of once.

As I referred to in another issue comment ([Ref](https://github.com/desktop/desktop/issues/5219#issuecomment-1858996516)), I would like to be able to use the existing 'amend commit' functionality on commits other than the latest. We can already reorder/squash commits from anywhere in the history; so it would seem that this would just be a UX improvement, and not require much additional underlying functionality/new concepts; and it sticks with the "use interactive rebase for specific features" aspect mentioned above. ([Ref](https://github.com/desktop/desktop/issues/12354#issuecomment-853810896))

---

RE: https://github.com/desktop/desktop/issues/12354#issuecomment-857513302

> We also added this in #12353 but only applies to the last commit as of today 😅

Are there any plans for this to be extended to any commit? (as I mentioned above)

---

RE:

> I think the only remaining request to get it down to zero unnecessary conflict resolutions would be the ability to edit a particular commit directly, meaning to temporarily return the branch to a particular prior state where one could make modifications, use the amend functionality you already offer, and then continue back to the present.

Please!

> As an added benefit, this would allow people to examine the state of the branch at a previous commit regardless of whether they desired to change it. I imagine this would be particularly desirable for a novice, who may find reading a diff very challenging initially.


This can be done with 'Checkout Commit' can't it?

![image](https://github.com/desktop/desktop/assets/753891/8229c133-5627-4779-ab29-a5ede4b62d9e)

---

RE: https://github.com/desktop/desktop/issues/12354#issuecomment-874366125

> Being able to change a commit message without squashing

See also:

- https://github.com/desktop/desktop/issues/5219

---

RE: https://github.com/desktop/desktop/issues/12354#issuecomment-875198692

> To reword a commit, presently you would have to reorder them (e.g., via drag and drop), amend it, and then reorder them back.

While it's good to know this is a workaround, it has the potential of introducing a bunch of unnecessary conflicts from reordering. A better workaround would be to do this on the command line: `git rebase -i <commit>` and then use `reword`

---

RE: https://github.com/desktop/desktop/issues/12354#issuecomment-882400092

> I'm afraid right now those features are not in our roadmap. I recommend to file issues for both features (if they don't exist already). Getting traction on those (via 👍 reactions or comments) will make them more likely to happen, as that would mean there is an interest/need on them.

Any chance of getting the `help-wanted` label added so the community can add these features?

---

See also:

- https://github.com/desktop/desktop/issues/10460

**Comment by 0xdevalias** [comment]
Dove into the source code for a bit of a search, and here are some of the lib files related to `squash` and `rebaseInteractive`. There may be more than this, and I didn't look too deeply as to where/how these are used by the UI itself, but it could be a good starting place for anyone wanting to look into this stuff deeper:

> From a quick search in the code, I found this git lib function for `squash` that sounds like it's probably used for the squash functionality (via interactive rebase):
> 
> https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/squash.ts#L10-L40
> 
> It then calls into `rebaseInteractive`:
> 
> https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/squash.ts#L148-L156
> 
> `rebaseInteractive` is defined here:
> 
> https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/rebase.ts#L533-L556
> 
> That seems to make the actual call to `git rebase -i` here, and doesn't look like it currently uses `--autostash` at all:
> 
> https://github.com/desktop/desktop/blob/6aa1e092c738da2851624fdf376daa658ccdaa69/app/src/lib/git/rebase.ts#L582-L594
> 
> _Originally posted by @0xdevalias in https://github.com/desktop/desktop/issues/12759#issuecomment-1859055247_

**Comment by PaddleStroke** [comment]
So I'm trying out github desktop (I was using git gui so far).
And the missing rebase interactive is a show stopper for me as I use it all the time.

It is not included in git gui neither, but there you can add custom commands.

**Comment by PaddleStroke** [comment]
So I have just seen https://github.com/desktop/desktop/pull/12235 or https://github.com/desktop/desktop/issues/12299 which are pretty nice replacements for rebase interactive indeed.
However I think that enabling at least custom actions is still a basic feature that would be good to have.

**Comment by Andrej730** [comment]
I think there could be a very simple introduction to interactive rebase in Github Desktop. 

E.g. by adding to commit's context menu new option "Start interactive rebase for this commit", which would be an equivalent for running `git rebase -i xxxxx` and would just start a text editor to continue the operation (e.g. I use VS Code + GitLens extenstion, which makes this process very easy). The only other thing needed from Github Desktop for this is to show popup with message "Interactive rebase is in progress" and button "Abort Rebase". This would be convenient and would help skip a few more cases when user would go to command line.

![image](https://github.com/desktop/desktop/assets/9417531/67648a8a-b20c-40af-9051-00bceebbaa9e)

**Comment by yzababababab** [comment]
Ok, probably is canceled 😕

Andrej ***@***.***> 於 2024年6月8日 週六 15:21 寫道：

**Comment by yzababababab** [comment]
Waiting 😔😦

妍妍yenyen ***@***.***> 於 2024年6月8日 週六 15:25 寫道：

**Comment by yzababababab** [comment]
Camino

妍妍yenyen ***@***.***> 於 2024年6月8日 週六 15:26 寫道：

**Comment by yzababababab** [comment]
Getting sign

妍妍yenyen ***@***.***> 於 2024年6月8日 週六 15:27 寫道：

**Comment by yzababababab** [comment]
妍妍yenyen ***@***.***> 於 2024年6月8日 週六 15:27 寫道：

**Comment by niik** [comment]
Hi folks, thanks so much for all the discussions and thoughts in this issue over the years. We discussed this issue at our most recent issue review meeting and we've decided against implementing a generic interactive rebase interface as was most recently suggested but since this issue was opened we have added support for reordering (dragging and dropping commits in the history view), cherry-picking (drag one or more commits from the history commit list to the branch dropdown and pick an existing or new branch to cherry pick them onto), squashing (drag one or more commits onto another commit) and amended commits (right click last commit in history, select amend commit).

Although we have no timeline for it we do want to expand this to support things like changing the commit message of an arbitrary commit but we won't be implementing a 1-1 mapping to the interactive rebase interface and commands of Git.
